Match file type on read

jedwards4b opened the pull request:
### Description of changes
Change for read_darray into variables of type logical.  Type in pio_read_darray should match type on file. 

### Specific notes

Contributors other than yourself, if any:

CTSM Issues Fixed (include github issue #): #1091 

Are answers expected to change (and if so in what way)?

Any User Interface Changes (namelist or namelist defaults changes)?

Testing performed, if any: SMS_D.f19_g17.IHistClm50Sp.cheyenne_intel with namelist changes noted in #1091 

(Documentation on system testing is here: https://github.com/ESCOMP/ctsm/wiki/System-Testing-Guide)
(aux_clm on cheyenne for intel/gnu and izumi for intel/gnu/nag/pgi is the standard for tags on master)

**NOTE: Be sure to check your coding style against the standard
(https://github.com/ESCOMP/ctsm/wiki/CTSM-coding-guidelines) and review
the list of common problems to watch out for
(https://github.com/ESCOMP/CTSM/wiki/List-of-common-problems).**

billsacks reviewed (changes requested):
@jedwards4b thank you for working on this. However, it looks like your fix here only solves the problem in the specific case where a double on the file is being read into a logical in memory. I don't see how this would have solved the original issue, where the problem arose from reading a double on file into an integer in memory. @swensosc also reported similar problems when reading a single precision on file into a double in memory.

So I think that some more general changes are needed here. You made a number of changes in 1d9bd3e7bdc59df96f2f4eef228658afb0d582dd related to changing how type conversion happens, and my initial thought is that all of them will need the same fix – or that we should just back out those changes for now.

I will also reiterate: If you are confident that the code will work right with PIO2, then we could leave this as is and add some error checking that prevents you from trying to do these data conversions if using PIO1, if you feel that adding this error checking would be easier than fixing the problem.

jedwards4b commented:
@billsacks I encouraged @swensosc  to open an issue - I haven't seen it.  I would be happy to look at another issue if it exists.

billsacks commented:
> @billsacks I encouraged @swensosc to open an issue - I haven't seen it. I would be happy to look at another issue if it exists.

My sense is that the problem @swensosc ran into is fundamentally the same as the one in #1091 when we tried to read a double into an integer. My sense is that this all comes down to the changes in 1d9bd3e7bdc59df96f2f4eef228658afb0d582dd. Maybe, before we spend a lot more time on this, it would be helpful to understand whether it would be reasonable to back out the relevant changes from that commit: what would be the downsides of doing so? (I know that some of the changes in that commit were needed for pio2, if I remember correctly, so I'm just referring to backing out the changes that were related to changing how type conversion happens.)

jedwards4b commented:
@billsacks I have refactored the rest of the pio_read_darray cases.

billsacks commented on src/main/ncdio_pio.F90.in:
```diff
@@ -1902,18 +1940,40 @@ contains
              if (present(nt)) then
                 call pio_setframe(ncid, vardesc, int(nt,kind=Pio_Offset_Kind))
              end if
+#if ({ITYPE}==TYPEINT)
+             if(xtype == PIO_INT .or. xtype == PIO_REAL) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+             elseif(xtype == PIO_DOUBLE) then
+                allocate(ddata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, ddata, status)
+                data = int(ddata)
+             endif
+#endif
+#if ({ITYPE}==TYPEDOUBLE)
+             if(xtype == PIO_DOUBLE) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+             else if(xtype == PIO_INT) then
+                allocate(idata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, idata, status)
+                data = real(idata, kind=R8)
+                deallocate(idata)
+             else if(xtype == PIO_REAL) then
+                allocate(fdata(size(data,1),size(data,2)))
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, fdata, status)
+                data = real(fdata, kind=R8)
+                deallocate(fdata)
+             endif
+#endif
```
- [ ] I'm having trouble seeing how this block of code works as it currently is if switchdim is given: The code that was in place before called `pio_read_darray` to read into `temp`, which is the transpose of `data`. So it feels like it shouldn't work to read directly into data, as is now being done. Is there something that I'm not understanding that allows this to work as it is now?

billsacks commented on src/main/ncdio_pio.F90.in:
```diff
@@ -1693,16 +1695,49 @@ contains
-             call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
+#if ({ITYPE}==TYPEINT)
+             if (xtype == PIO_INT .or. xtype == PIO_REAL) then
+                call pio_read_darray(ncid, vardesc, iodesc_plus%iodesc, data, status)
```
- [ ] So, to confirm: it's always safe to read a single-precision real on the file into an integer in memory?

billsacks reviewed (changes requested):
Thank you very much for continuing to work on this, @jedwards4b . I have two inline questions below. In addition:

- [ ] Can you briefly describe the advantages of this approach compared with what was in place before 1d9bd3e7bdc59df96f2f4eef228658afb0d582dd? There is more code and complexity now, so I want to be sure that it is justified.

- [ ] It looks like you have addressed all of the read places that were changed in 1d9bd3e7bdc59df96f2f4eef228658afb0d582dd (thank you). But the writes that were changed in 1d9bd3e7bdc59df96f2f4eef228658afb0d582dd had no changes here. Can you confirm that no changes are needed when writing?

- [ ] Can you please summarize what testing you have done, if any, to verify that this code works, e.g., by reading a variable of one type on the file into a different type in memory?

By the way: we use check boxes in CTSM PRs to track the status of reviewer requests. I (as the reviewer) will check these off once they have been addressed.

billsacks commented:
@jedwards4b thank you for your continued work on this. Before going further, I have a big picture question: It feels like the new code you have added is very generic code that would presumably be needed in other components that use PIO. As generic utility code, it feels like this belongs in PIO itself – not requiring every component to implement and maintain code like this. Can you describe the reason for putting this in CTSM rather than in PIO?

jedwards4b commented:
The original problem was the conversion of type double to logical - this is not a conversion that I think can be done in a more general implementation because how this conversion is done is unique to the end case - for example is 1.0 true or is any non 0.0 value true?  I fixed that issue and the feedback was that I was not handling a conversion of int to double, so then I fixed that case, then you correctly pointed out that I did not handle the switchdim correctly, so now I've fixed that and you tell me that now the solution is too general.

billsacks commented:
> now I've fixed that and you tell me that now the solution is too general.

Sorry, I did not mean to say that your solution is too general. Indeed, what is needed here *is* a general solution, and I'm sorry if my wording suggested that I was critical of this general-purpose code: that's not at all what I meant.

What I meant by my question is: Given that what's needed is a general solution to a general problem, would it make more sense to have the implementation in the pio layer so that all pio users could take advantage of this solution? Or is there some reason why it doesn't make sense to do that? I'm not trying to criticize your solution here, and I am truly appreciative of all of the time you have spent working on this. I just want to make sure that the solution we're putting in place is indeed the one that makes the most sense.

billsacks commented:
This PR is superseded by #1189 , which does a bit of refactoring and adds a bunch of tests. So I'm closing this PR.
